File: src/main/java/ruter/realtime/RuterService.java
package ruter.realtime;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.HttpClientBuilder;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;

public class RuterService {

    private final ObjectMapper mapper;
	private final HttpClient httpClient;

	public RuterService() {
        mapper = new ObjectMapper();
        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        mapper.setPropertyNamingStrategy(new PropertyNamingStrategy.UpperCamelCaseStrategy());

		httpClient = HttpClientBuilder.create()
                .build();
	}

    public UpcomingDepartureToDowntown fetchRealtimeInformation() {
        try {
            HttpResponse response = httpClient.execute(new HttpGet("http://reis.trafikanten.no/reisrest/realtime/getrealtimedata/3012130"));

            try (InputStream inputStream = response.getEntity().getContent()) {
                List<BusDepartureDto> departureDtos = mapper.readValue(inputStream, new TypeReference<List<BusDepartureDto>>() { });

                return new UpcomingDepartureToDowntown(departureDtos);
            }
        }
        catch (IOException e) {
            throw new RuterIOException(e);
        }
    }

}

File: src/main/kotlin/ruter/realtime/RuterService.kt
package ruter.realtime

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategy
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import org.apache.http.client.HttpClient
import org.apache.http.client.methods.HttpGet
import org.apache.http.impl.client.HttpClientBuilder

import java.io.IOException
import java.util.stream.Collectors

open class RuterService {

    private val mapper: ObjectMapper = ObjectMapper()
    private val httpClient: HttpClient

    init {
        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                .registerKotlinModule()
                .registerModule(JavaTimeModule())

        mapper.propertyNamingStrategy = PropertyNamingStrategy.UpperCamelCaseStrategy()

        httpClient = HttpClientBuilder.create()
                .build()
    }

    fun fetchRealtimeInformation(): UpcomingDeparture? {
        return fetchRealtime(RealTimeRequest()).departures.firstOrNull()
    }

    open fun fetchRealtime(request: RealTimeRequest): UpcomingDepartures {
        try {
            val response = httpClient.execute(HttpGet("http://reisapi.ruter.no/StopVisit/GetDepartures/${request.stopId}"))

            response.entity.content.use { inputStream ->
                val departureDtos = mapper.readValue<List<DepartureDto>>(inputStream, object : TypeReference<List<DepartureDto>>() {})

                return UpcomingDepartures(request, departureDtos)
            }
        } catch (e: IOException) {
            throw RuterIOException(e)
        }

    }

}
